src/translator/nth_element.cpp

```diff
@@ -59,23 +31,31 @@ class NthElementCPU {
     const auto inputN    = scores->shape()[-2];
     const auto dimBatch  = scores->shape()[-4];
     ABORT_IF(inputN != (isFirst ? 1 : N), "Input tensor has wrong beam dim??"); // @TODO: Remove isFirst argument altogether
-
-    std::vector<int> cumulativeBeamSizes(dimBatch + 1, 0);
-    std::vector<int> batchFirstElementIdxs(dimBatch + 1, 0);
-
-    for(int batchIdx = 0; batchIdx < dimBatch; ++batchIdx) {
-      cumulativeBeamSizes[batchIdx + 1] = (batchIdx + 1) * (int)N;
-      batchFirstElementIdxs[batchIdx + 1] += (batchIdx + 1) * inputN * vocabSize;
-      ABORT_IF(cumulativeBeamSizes[batchIdx + 1] != cumulativeBeamSizes[batchIdx] + (int)N, "cumulativeBeamSizes wrong??");
-      ABORT_IF((isFirst ? batchIdx + 1 : cumulativeBeamSizes[batchIdx + 1]) != (batchIdx + 1) * inputN, "inputN wrong??");
-    }
-    ABORT_IF(cumulativeBeamSizes.back() != dimBatch * N, "cumulativeBeamSizes.back() wrong??");
+    const float* scoresData = scores->data();
 
     size_t maxSize = N * dimBatch;
     h_res.resize(maxSize);
     h_res_idx.resize(maxSize);
 
-    selectNBest(scores->data(), batchFirstElementIdxs, cumulativeBeamSizes);
+    std::vector<int> idxs(dimBatch * inputN * vocabSize);
+    std::iota(idxs.begin(), idxs.end(), 0);
```

Comment:
I know this was in the pre-existing code, but I think it can be optimized further.  
```
std::vector<int> idxs(inputN * vocabSize);
std::iota(idxs.begin(), idxs.end(), 0);
```
Then in the loop reuse the same indices for each batch.  
```
const float *batchScores = scoresData + batchIdx * inputN * vocabSize;
std::partial_sort(idxs.begin(), idxs.begin() + N, idxs.end(), [&](int a, int b) { return batchScores[a] > batchScores[b]; }
```
If hypotheses have ties this could make tie-breaking batch dependent but it wasn't guaranteed to be anything consistent anyway.  
```